lib/common/options.c

```diff
@@ -939,68 +939,80 @@ pcmk__cluster_option(GHashTable *options, const char *name)
  * \internal
  * \brief Add a description element to a meta-data string
  *
- * \param[in,out] s       Meta-data string to add to
+ * \param[in,out] parent  Parent XML node
  * \param[in]     tag     Name of element to add (\c PCMK_XE_LONGDESC or
  *                        \c PCMK_XE_SHORTDESC)
  * \param[in]     desc    Textual description to add
  * \param[in]     values  If not \p NULL, the allowed values for the parameter
- * \param[in]     spaces  If not \p NULL, spaces to insert at the beginning of
- *                        each line
+ *
+ * \return Standard Pacemaker return code
  */
-static void
-add_desc(GString *s, const char *tag, const char *desc, const char *values,
-         const char *spaces)
+static int
+add_desc(xmlNode *parent, const char *tag, const char *desc, const char *values)
 {
-    char *escaped_en = crm_xml_escape(desc);
-
-    if (spaces != NULL) {
-        g_string_append(s, spaces);
-    }
-    pcmk__g_strcat(s,
-                   "<", tag, " " PCMK_XA_LANG "=\"" PCMK__VALUE_EN "\">",
-                   escaped_en, NULL);
+    const char *full_desc = desc;
+    xmlNode *node = NULL;
+    char *desc_with_values = NULL;
+    int rc = pcmk_rc_ok;
 
     if (values != NULL) {
-        // Append a period if desc doesn't end in "." or ".)"
-        if (!pcmk__str_empty(escaped_en)
-            && (s->str[s->len - 1] != '.')
-            && ((s->str[s->len - 2] != '.') || (s->str[s->len - 1] != ')'))) {
+        // @TODO Use XML node for allowed values
+        bool need_period = !pcmk__ends_with(desc, ".")
+                           && !pcmk__ends_with(desc, ".)");
+
+        desc_with_values = crm_strdup_printf("%s%s Allowed values: %s.",
+                                             desc, (need_period? "." : ""),
+                                             values);
+        full_desc = desc_with_values;
+    }
 
-            g_string_append_c(s, '.');
-        }
-        pcmk__g_strcat(s, " Allowed values: ", values, NULL);
-        g_string_append_c(s, '.');
+    node = pcmk_create_xml_text_node(parent, tag, full_desc);
+    if (node == NULL) {
+        rc = ENOMEM;
+        goto done;
     }
-    pcmk__g_strcat(s, "</", tag, ">\n", NULL);
+    crm_xml_add(node, PCMK_XA_LANG, PCMK__VALUE_EN);
 
 #ifdef ENABLE_NLS
     {
         static const char *locale = NULL;
+        const char *local_desc = _(desc);
 
-        char *localized = crm_xml_escape(_(desc));
```

Comment:
I assume this is why the regression test output is no longer escaped.  I'd also keep this since who knows what changing it will break.

Incidentally, there's a comment in crm_xml_escape from 2013 about how xmlNodeDump is slow.  I wonder if that's still true (or was ever true) and if we therefore even need to keep doing our own escaping instead of asking libxml to do it for us.

Reply:
> I'd also keep this since who knows what changing it will break.

In general, we don't escape the contents of XML text nodes. See `pcmk__xml2text()` -> `dump_xml_text()`. Our high-level dump functions all call `pcmk__xml2text()`. That's what makes me feel a lot more comfortable about this.

There are a few exceptions. We're not really consistent, so it's not clear to me why we escape the output in some certain cases. Sometimes we do our own escaping, sometimes not. ometimes we build a string manually, sometimes we tell libxml to dump it for us (which caused some buffering issues; see next major point below). I'm not going to blindly assume that we HAVE a good reason for everything. We do escape the following:
* XML attribute values -- see `pcmk__dump_xml_attr()` and `show_xml_element()`
* the exit reason in the XML version of the `"failed-action"` message
  * This becomes an XML attribute value. Until c63a898b, it would have been escaped by the "XML attribute values" case, so this would've been entirely redundant. However, with that commit, we started using `pcmk__xml2fd()` -> `xmlNodeDumpOutput()` in `lib/common/output_xml.c`, like poki wanted to do. The libxml dump functions don't escape all the special characters that `crm_xml_escape()` does.
* the description of a systemd unit in its text OCF metadata (see `systemd_unit_metadata()`)

Everything else goes unescaped.

> Incidentally, there's a comment in crm_xml_escape from 2013 about how xmlNodeDump is slow. I wonder if that's still true (or was ever true) and if we therefore even need to keep doing our own escaping instead of asking libxml to do it for us.

Yeah... that comment saying `xmlNodeDump()` is too slow is from beekhof. Then poki drafted c63a898b as mentioned above, which started using `xmlNodeDumpOutput()` (the modern version of `xmlNodeDump()`). His argument was that using `xmlNodeDumpOutput()` would make things faster. So who even knows?

1382915a (from you) reverts most of poki's commit in order to fix a regression that it caused. The only place where we still use `pcmk__xml2fd()` is when finishing XML formatted output.

Reply:
@kgaillot tag as an FYI on these inconsistencies, even though I don't think they're relevant here since this is consistent with our current practices for text nodes. You might already be aware of all this.

Reply:
I'm pretty sure it does need to be escaped. If someone puts a `<` in the description, it will break the XML if not escaped. If we're not escaping text nodes elsewhere, either that's a bug or we know the content doesn't need escaping.

Reply:
> I'm pretty sure it does need to be escaped. If someone puts a `<` in the description, it will break the XML if not escaped.

Confirmed. Pacemaker fails to build if you put that into one of the options. I added `<test>` to the content of a longdesc and `make` failed with
```
/home/reid/git/pacemaker/daemons/controld/pacemaker-controld metadata > pacemaker-controld.xml
/usr/bin/xsltproc --nonet --novalid --stringparam man.name pacemaker-controld \
	 ../../xml/ocf-meta2man.xsl \
	/home/reid/git/pacemaker/daemons/controld/pacemaker-controld.xml > /home/reid/git/pacemaker/daemons/controld/pacemaker-controld.dbook
/home/reid/git/pacemaker/daemons/controld/pacemaker-controld.xml:18: parser error : Opening and ending tag mismatch: test line 17 and shortdesc
      </shortdesc>
...
```

> If we're not escaping text nodes elsewhere, either that's a bug or we know the content doesn't need escaping.

I'd say it's a bug then.
* `dump_xml_text()` does our general-purpose XML-to-text dumping for text nodes, and it doesn't escape text node contents. We could say we're assuming the text got escaped when we added it to the node. However, we escape XML attribute values in `pcmk__dump_xml_attr()` -- so we don't assume it for attribute values.
* `show_xml_node()`/`show_xml_element()` (the formatted output display functions) appear to skip text nodes entirely.
* systemd metadata *does* get `crm_xml_escape()`'d.
* LSB and st_lha metadata... might? They call `xmlEncodeEntitiesReentrant()`... which at a glance looks like a libxml-provided equivalent of `crm_xml_escape()`. That's probably worth looking into further to compare and contrast.

I have no idea about the state of `xmlNodeDumpOutput()` as discussed above... conflicting statements, obfuscated comments and commit messages, and a history of unexpected behavior (see the buffering issue).

Reply:
https://projects.clusterlabs.org/T768